**Here are some instructions you need to follow to obtain the correct answers:**

- Please remove the column containing the name of each university 
- Obtain descriptive statistics on the numeric columns of the dataset using the .describe() method from Pandas 
    - Round these statistics to 2 decimal places using the .round() method from Pandas 
- Consider ‘Grad.Rate’ as the target variable 
- Do not normalise or scale the data in any way 
- Do not forget to one-hot encode the ‘Private’ variable 
    - You can achieve this using pandas.get_dummies() 
    - Be sure to set drop_first = True so the redundant encoded column is excluded 
- When splitting the data, please use sklearn.model_selection.train_test_split() 
    - Please use a 70:30 train:test split 
    - Please set random_state = 42 so that the results you obtain can be compared to ours 
- Fit the model on the training set 
- Output the model coefficients, rounded to 2 decimal places using the .round() method from Pandas 
- Obtain predictions on both, the training and testing sets 
- Generate the following metrics (using sklearn.metrics) on both, the training and testing sets, and round them to 2 decimal places using the round() function 
    - R² 
    - Mean Squared Error 
    - Mean Absolute Error 

| Column          | Meaning                                                                                 |
| --------------- | --------------------------------------------------------------------------------------- |
| **Unnamed: 0**  | The college name (currently just an index column in your file).                         |
| **Private**     | Factor indicating whether the college is **private** or **public**.                     |
| **Apps**        | Number of applications received.                                                        |
| **Accept**      | Number of applicants accepted.                                                          |
| **Enroll**      | Number of new students enrolled.                                                        |
| **Top10perc**   | Percentage of new students who graduated in the **top 10% of their high school class**. |
| **Top25perc**   | Percentage of new students who graduated in the **top 25% of their high school class**. |
| **F.Undergrad** | Number of **full-time undergraduates**.                                                 |
| **P.Undergrad** | Number of **part-time undergraduates**.                                                 |
| **Outstate**    | **Out-of-state tuition** cost.                                                          |
| **Room.Board**  | Cost of **room and board** (housing + meals).                                           |
| **Books**       | Estimated cost of **books**.                                                            |
| **Personal**    | Estimated **personal spending** by students.                                            |
| **PhD**         | Percentage of faculty with a **Ph.D. degree**.                                          |
| **Terminal**    | Percentage of faculty with a **terminal degree** (highest degree in their field).       |
| **S.F.Ratio**   | **Student-to-faculty ratio**.                                                           |
| **perc.alumni** | Percentage of alumni who **donate**.                                                    |
| **Expend**      | Instructional **expenditure per student**.                                              |
| **Grad.Rate**   | **Graduation rate** (% of students who graduate).                                       |


In [ ]:
# Import required libraries
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error, mean_absolute_percentage_error
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

#!conda install openpyxl -y

In [62]:
data_path = '/Users/aditikulkarni/Documents/Masters/AI-Projects/04-ML-Models/linear-regression/data/raw/'  # Path to the dataset
df = pd.read_excel(data_path + 'College.xlsx', header=0)
df.head()

Unnamed: 0 Private  Apps  Accept  Enroll  Top10perc  \
0  Abilene Christian University     Yes  1660    1232     721         23   
1            Adelphi University     Yes  2186    1924     512         16   
2                Adrian College     Yes  1428    1097     336         22   
3           Agnes Scott College     Yes   417     349     137         60   
4     Alaska Pacific University     Yes   193     146      55         16   

   Top25perc  F.Undergrad  P.Undergrad  Outstate  Room.Board  Books  Personal  \
0         52         2885          537      7440        3300    450      2200   
1         29         2683         1227     12280        6450    750      1500   
2         50         1036           99     11250        3750    400      1165   
3         89          510           63     12960        5450    450       875   
4         44          249          869      7560        4120    800      1500   

   PhD  Terminal  S.F.Ratio  perc.alumni  Expend  Grad.Rate  
0   70        78       18.1           12    7041         60  
1   29        30       12.2           16   10527         56  
2   53        66       12.9           30    8735         54  
3   92        97        7.7           37   19016         59  
4   76        72       11.9            2   10922         15

In [63]:
print (df.shape)
print (df.info())

(777, 19)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 777 entries, 0 to 776
Data columns (total 19 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Unnamed: 0   777 non-null    object 
 1   Private      777 non-null    object 
 2   Apps         777 non-null    int64  
 3   Accept       777 non-null    int64  
 4   Enroll       777 non-null    int64  
 5   Top10perc    777 non-null    int64  
 6   Top25perc    777 non-null    int64  
 7   F.Undergrad  777 non-null    int64  
 8   P.Undergrad  777 non-null    int64  
 9   Outstate     777 non-null    int64  
 10  Room.Board   777 non-null    int64  
 11  Books        777 non-null    int64  
 12  Personal     777 non-null    int64  
 13  PhD          777 non-null    int64  
 14  Terminal     777 non-null    int64  
 15  S.F.Ratio    777 non-null    float64
 16  perc.alumni  777 non-null    int64  
 17  Expend       777 non-null    int64  
 18  Grad.Rate    777 non-null    int64  
dty

In [64]:
df.describe().transpose().round(2)

count      mean      std     min     25%     50%      75%  \
Apps         777.0   3001.64  3870.20    81.0   776.0  1558.0   3624.0   
Accept       777.0   2018.80  2451.11    72.0   604.0  1110.0   2424.0   
Enroll       777.0    779.97   929.18    35.0   242.0   434.0    902.0   
Top10perc    777.0     27.56    17.64     1.0    15.0    23.0     35.0   
Top25perc    777.0     55.80    19.80     9.0    41.0    54.0     69.0   
F.Undergrad  777.0   3699.91  4850.42   139.0   992.0  1707.0   4005.0   
P.Undergrad  777.0    855.30  1522.43     1.0    95.0   353.0    967.0   
Outstate     777.0  10440.67  4023.02  2340.0  7320.0  9990.0  12925.0   
Room.Board   777.0   4357.53  1096.70  1780.0  3597.0  4200.0   5050.0   
Books        777.0    549.38   165.11    96.0   470.0   500.0    600.0   
Personal     777.0   1340.64   677.07   250.0   850.0  1200.0   1700.0   
PhD          777.0     72.66    16.33     8.0    62.0    75.0     85.0   
Terminal     777.0     79.70    14.72    24.0    71.0    82.0     92.0   
S.F.Ratio    777.0     14.09     3.96     2.5    11.5    13.6     16.5   
perc.alumni  777.0     22.74    12.39     0.0    13.0    21.0     31.0   
Expend       777.0   9660.17  5221.77  3186.0  6751.0  8377.0  10830.0   
Grad.Rate    777.0     65.46    17.18    10.0    53.0    65.0     78.0   

                 max  
Apps         48094.0  
Accept       26330.0  
Enroll        6392.0  
Top10perc       96.0  
Top25perc      100.0  
F.Undergrad  31643.0  
P.Undergrad  21836.0  
Outstate     21700.0  
Room.Board    8124.0  
Books         2340.0  
Personal      6800.0  
PhD            103.0  
Terminal       100.0  
S.F.Ratio       39.8  
perc.alumni     64.0  
Expend       56233.0  
Grad.Rate      118.0

In [65]:
df.isnull().sum()
df.drop(columns=['Unnamed: 0'], inplace=True)
df.dropna(inplace=True)  # Drop rows with missing values

In [50]:
#df['Private'].value_counts()

# For column Private, map Yes to 1 and No to 0

#df['Private'] = df['Private'].map({"Yes": 1, "No": 0})
#df['Private'].value_counts()


In [66]:
# One-hot encode 'Private'
df = pd.get_dummies(df, columns=['Private'], drop_first=True)
print (df.head())
print (df.shape)

   Apps  Accept  Enroll  Top10perc  Top25perc  F.Undergrad  P.Undergrad  \
0  1660    1232     721         23         52         2885          537   
1  2186    1924     512         16         29         2683         1227   
2  1428    1097     336         22         50         1036           99   
3   417     349     137         60         89          510           63   
4   193     146      55         16         44          249          869   

   Outstate  Room.Board  Books  Personal  PhD  Terminal  S.F.Ratio  \
0      7440        3300    450      2200   70        78       18.1   
1     12280        6450    750      1500   29        30       12.2   
2     11250        3750    400      1165   53        66       12.9   
3     12960        5450    450       875   92        97        7.7   
4      7560        4120    800      1500   76        72       11.9   

   perc.alumni  Expend  Grad.Rate  Private_Yes  
0           12    7041         60         True  
1           16   10527        

In [75]:
predictors = ['Apps','Accept','Enroll', 'Private','Top10perc','Top25perc','F.Undergrad','P.Undergrad','Outstate','Room.Board','Books','Personal','PhD','Terminal','S.F.Ratio','perc.alumni','Expend']
target = ['Grad.Rate']

#X = df[predictors]
X = df.drop(columns=target)
y = df[target]

In [73]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
print (X_train.shape, X_test.shape, y_train.shape, y_test.shape)

model = LinearRegression()
model.fit(X_train, y_train)
print(model.intercept_)

(543, 17) (234, 17) (543,) (234,)
34.93871250622203


In [69]:
y_pred = model.predict(X_test)
y_train_pred = model.predict(X_train)

In [70]:
def regression_metrics(y_pred, y_test):
    r2 = r2_score(y_test, y_pred)
    mse = mean_squared_error(y_test, y_pred)
    rmse = np.sqrt(mse)
    mae = mean_absolute_error(y_test, y_pred)
    mape = mean_absolute_percentage_error(y_test, y_pred) * 100

    var = y_test.var().item()
    std = y_test.std().item()
    mean_val = y_test.mean().item()

    return r2, mse, rmse, mae, mape, var, std, mean_val

metrics_train = regression_metrics(y_train_pred, y_train)
metrics_test = regression_metrics(y_pred, y_test)

metrics_df = pd.DataFrame({
    'Dataset' : ['Training', 'Testing'],
    'R2' : [metrics_train[0], metrics_test[0]],
    'MSE' : [metrics_train[1], metrics_test[1]],
    'Variance' : [metrics_train[5], metrics_test[5]],
    'RMSE' : [metrics_train[2], metrics_test[2]],
    'Std Dev' : [metrics_train[6], metrics_test[6]],
    'MAE' : [metrics_train[3], metrics_test[3]],
    'Mean' : [metrics_train[7], metrics_test[7]],
    'MAPE' : [metrics_train[4], metrics_test[4]],
}).set_index('Dataset')

metrics_df.round(2)

R2     MSE  Variance   RMSE  Std Dev   MAE   Mean   MAPE
Dataset                                                             
Training  0.46  160.59    299.49  12.67    17.31  9.37  65.20  17.28
Testing   0.44  160.48    285.52  12.67    16.90  9.60  66.08  18.04

Here are some thumb rules when it comes to evaluating our metrics

| Metric | Thumb Rule / Interpretation |
|--------|----------------------------|
| MSE | RMSE lower than variance means model improves over naïve prediction |
| RMSE | MSE lower than standard deviation indicates good fit |
| MAE | Compare to typical value (e.g., the mean) of target: roughly tells typical error magnitude; lower is better |
| MAPE | Interprets error as % of true value; smaller % is better |
| $R^2$ | Proportion of variance explained: closer to 1 is better; 0 means no improvement over mean |



In [72]:
# Coefficients
coef_df = pd.DataFrame(model.coef_, index=X.columns, columns=['Coefficient']).round(2)
print(coef_df)


             Coefficient
Apps                0.00
Accept             -0.00
Enroll              0.00
Top10perc           0.03
Top25perc           0.14
F.Undergrad        -0.00
P.Undergrad        -0.00
Outstate            0.00
Room.Board          0.00
Books              -0.00
Personal           -0.00
PhD                 0.14
Terminal           -0.11
S.F.Ratio          -0.03
perc.alumni         0.34
Expend             -0.00
Private_Yes         5.05


In [71]:
# Drop college names
df = pd.read_excel(data_path + 'College.xlsx', header=0)
df.drop(columns=['Unnamed: 0'], inplace=True)

# Descriptive stats
print(df.describe().round(2))

# One-hot encode 'Private'
df = pd.get_dummies(df, columns=['Private'], drop_first=True)

# Split predictors and target
X = df.drop(columns=['Grad.Rate'])
y = df['Grad.Rate']

# Train/test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Fit model
model = LinearRegression()
model.fit(X_train, y_train)

# Predictions
y_train_pred = model.predict(X_train)
y_test_pred = model.predict(X_test)

# Metrics
def regression_metrics(y_true, y_pred):
    return {
        'R2': round(r2_score(y_true, y_pred), 2),
        'MSE': round(mean_squared_error(y_true, y_pred), 2),
        'MAE': round(mean_absolute_error(y_true, y_pred), 2)
    }

print("Training:", regression_metrics(y_train, y_train_pred))
print("Testing:", regression_metrics(y_test, y_test_pred))

# Coefficients
coef_df = pd.DataFrame(model.coef_, index=X.columns, columns=['Coefficient']).round(2)
print(coef_df)


           Apps    Accept   Enroll  Top10perc  Top25perc  F.Undergrad  \
count    777.00    777.00   777.00     777.00      777.0       777.00   
mean    3001.64   2018.80   779.97      27.56       55.8      3699.91   
std     3870.20   2451.11   929.18      17.64       19.8      4850.42   
min       81.00     72.00    35.00       1.00        9.0       139.00   
25%      776.00    604.00   242.00      15.00       41.0       992.00   
50%     1558.00   1110.00   434.00      23.00       54.0      1707.00   
75%     3624.00   2424.00   902.00      35.00       69.0      4005.00   
max    48094.00  26330.00  6392.00      96.00      100.0     31643.00   

       P.Undergrad  Outstate  Room.Board    Books  Personal     PhD  Terminal  \
count       777.00    777.00      777.00   777.00    777.00  777.00    777.00   
mean        855.30  10440.67     4357.53   549.38   1340.64   72.66     79.70   
std        1522.43   4023.02     1096.70   165.11    677.07   16.33     14.72   
min           1.00

----

**1. What is the mean graduation rate in the provided dataset before the train:test split? Round your answer to 2 decimal places and pick the closest option from the following.**
- 70.23
- 65.46
- 40.00
- 35.23

**Answer:** 65.46

**2.What is the value of the intercept term that you obtain after fitting your model on the training data. Round your answer to 2 decimal places and pick the closest option from the following.**
- 34.94
- 0.13
- 100.13
- 0.94

**Answer:** 34.94

**3.Excluding categorical predictors, which predictor has the highest coefficient after fitting the model on the training data?** 
- ‘Enroll’
- ‘PhD’
- ‘perc.alumni’ 
- ‘Room.Board' 

**Answer:** perc.alumni

**4. What is the R² when predicting on the training data? Round your answer to 2 decimal places and pick the closest option from the following.**
- 0.89
- 0.72
- 0.46
- 0.39 

**Answer:** 0.46

**5. What is the mean squared error (MSE) when predicting on the testing data? Round your answer to 2 decimal places and pick the closest option from the following.**
- 1125.93
- 200.05
- 171.92
- 160.48

**Answer:** 160.48